Refactor GitHubAuth to prevent local configuration from causing test failures or harmful logging

ParkerM opened the pull request:
This PR adds a `CredentialAccessor` interface (any suggestions for a better name are very welcome) and moves the system calls from `GitHubAuth.kt` into a default implementation.

#### Reasoning
I created a token and assigned it to an env var in order to run the updater locally. The presence of the token var caused the auth tests to fail, and resulted in credentials being logged via assertion errors. This PR _should_ hopefully prevent this from happening if a scenario were to arise where a production token were inadvertently read by a test job in CI.

These changes move the `System.get[thing]` calls into a subclass that is supplied as a default constructor arg to the `GitHubAuth` class. The test suite has been updated to use a mock implementation to verify behavior, and existing functional tests have been moved into a block that is ignored if an environment variable or system property named `GITHUB_TOKEN` is present.

Previous failure message with `GITHUB_TOKEN` env var set:
```sh
GitHubAuthTest.readsTokenFromFile:20 expected:<f[oo]> but was:<f[my-real-token]>
```

A "successful" run of these tests with a token env var or system property set will now show up like so in the logs:
```sh
[INFO] Running net.adoptopenjdk.api.GitHubAuthTest
[WARNING] Tests run: 4, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 2.343 s - in net.adoptopenjdk.api.GitHubAuthTest
```

Other changes:
 * Use native temp directory in functional tests to improve OS-independence.
 * Restore `user.home` property if it's modified by a test.
 * Rename `GithubAuth` -> `GitHubAuth` to match convention of other classes.

# Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] You added tests to cover the change
- [x] `mvn clean install` build and test completes
- [ ] You changed or added to the documentation

ParkerM commented:
One thing I was unsure about was the expected behavior of this bit [here](https://github.com/AdoptOpenJDK/openjdk-api-v3/blob/ab1b8b154759f78a0c15db9a0f6bd60642e9b1a0/adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/dataSources/github/graphql/clients/GraphQLGitHubInterface.kt#L38-L40). From what I could tell according to the existing `readsTokenNullFromFile` test, if no token is set then an empty string is expected so I left that behavior in place, but it seems like that means the `"No token provided"` exception would never be thrown.

johnoliver commented on adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/dataSources/github/GitHubAuth.kt:
```diff
@@ -0,0 +1,29 @@
+        private val LOGGER = LoggerFactory.getLogger(this::class.java)
+    }
+
+    private val tokenName = "GITHUB_TOKEN"
```
I think I would move this to CredentialAccessor and make it the default argument

nickebbitt commented:
I think this PR improves what's already there in terms of handling the tokens and testability 👍 

Maybe not a conversation for this PR specifically but it's probably worth considering whether we need 3 different ways of providing the auth token? It seems like additional complexity in the code base for minimal value.

What's the current production mechanism for reading the auth token?

When it comes to creating cloud native 12 factor apps, the recommendation is to expose config via the environment so could that be the approach to consolidate on?

ParkerM commented:
> Maybe not a conversation for this PR specifically but it's probably worth considering whether we need 3 different ways of providing the auth token? It seems like additional complexity in the code base for minimal value.

+1

> When it comes to creating cloud native 12 factor apps, the recommendation is to expose config via the environment so could that be the approach to consolidate on?

It looks like Quarkus provides a convention for this: [Quarkus - Configuring Your Application](https://quarkus.io/guides/config)  
They also have a follow-on guide for [integrating with Spring Cloud Config](https://quarkus.io/guides/spring-cloud-config-client), which would help further decouple deployments from env-specific runtime properties.

nickebbitt commented:
> It looks like Quarkus provides a convention for this: [Quarkus - Configuring Your Application](https://quarkus.io/guides/config)
> They also have a follow-on guide for [integrating with Spring Cloud Config](https://quarkus.io/guides/spring-cloud-config-client), which would help further decouple deployments from env-specific runtime properties.

This will probably work for the `front-end` however the `updater` is not a Quarkus app. It's fairly vanilla Kotlin, the entry point being the [V3Updater](https://github.com/AdoptOpenJDK/openjdk-api-v3/blob/master/adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/V3Updater.kt#L29) class.

@johnoliver I wonder whether rather than discussing this here we should raise a separate issue. I feel there is a good discussion to be had around our approach to wiring (i.e. application start-up and dependency injection) across each component (i.e. front-end and updater) and how we reduce the coupling to the "infrastructure" code (i.e. database and network/http).

ParkerM commented:
Ah good catch. Agreed that the conventional approach should probably be part of a larger discussion.

I went ahead and reverted my changes and just updated the test to skip itself if the env var or system property is detected (which still ultimately addresses my concern). I also included the changes that restore any modified system properties and clean up temp files created by the test.  

I'm stumped as to why this would result in coverage to go down in an unrelated module. Could it be the case that restoring `user.home` in the test is reducing line hits later on? I did initially notice that the `readsTokenNullFromFile` test would fail on its own unless its sibling test ran before it, and with the addition of the before/after cleanup blocks that test actually returns null now instead of an empty string.
